tests/test_interp_utils.py

```diff
@@ -0,0 +1,327 @@
+    assert c.shape[1:] == x.shape[x.ndim - (c.ndim - 1) :]
+    assert np.unique((c.shape[0],) + x.shape[c.ndim - 1 :]).size == x.ndim - 1
+    test(x, c)
+
+
+class TestFastInterp:
+    """Test fast interpolation."""
+
+    @pytest.mark.unit
+    @pytest.mark.parametrize("N", [2, 6, 7])
+    def test_cheb_pts(self, N):
+        """Test we use Chebyshev points compatible with standard definition of DCT."""
+        np.testing.assert_allclose(cheb_pts(N), chebpts1(N)[::-1], atol=1e-15)
+        np.testing.assert_allclose(
+            cheb_pts(N, lobatto=True, domain=(-np.pi, np.pi)),
+            np.pi * chebpts2(N)[::-1],
+            atol=1e-15,
+        )
+
+    @pytest.mark.unit
+    @pytest.mark.parametrize("M", [1, 8, 9])
+    def test_rfftfreq(self, M):
+        """Make sure numpy uses Nyquist interpolant frequencies."""
+        np.testing.assert_allclose(np.fft.rfftfreq(M, d=1 / M), np.arange(M // 2 + 1))
+
+    @staticmethod
+    def _interp_rfft_harmonic(xq, f):
+        M = f.shape[-1]
+        fq = jnp.linalg.vecdot(
+            harmonic_vander(xq, M), harmonic(rfft(f, norm="forward"), M)
+        )
+        return fq
+
+    @staticmethod
+    def _f_1d(x):
+        """Test function for 1D FFT."""
+        return np.cos(7 * x) + np.sin(x) - 33.2
+
+    @staticmethod
+    def _f_1d_nyquist_freq():
+        return 7
+
+    @pytest.mark.unit
+    @pytest.mark.parametrize(
+        "func, n",
+        [
+            (_f_1d, 2 * _f_1d_nyquist_freq() + 1),
+            (_f_1d, 2 * _f_1d_nyquist_freq()),
+        ],
+    )
+    def test_interp_rfft(self, func, n):
+        """Test non-uniform FFT interpolation."""
+        xq = np.array([7.34, 1.10134, 2.28])
+        x = np.linspace(0, 2 * np.pi, n, endpoint=False)
+        assert not np.any(np.isclose(xq[..., np.newaxis], x))
+        f, fq = func(x), func(xq)
+        np.testing.assert_allclose(self._interp_rfft_harmonic(xq, f), fq)
+        np.testing.assert_allclose(interp_rfft(xq, f), fq)
+
+    @staticmethod
+    def _f_2d(x, y):
+        """Test function for 2D FFT."""
+        x_freq, y_freq = 3, 5
+        return (
+            # something that's not separable
+            np.cos(x_freq * x) * np.sin(2 * x + y)
+            + np.sin(y_freq * y) * np.cos(x + 3 * y)
+            # DC terms
+            - 33.2
+            + np.cos(x)
+            + np.cos(y)
+        )
+
+    @staticmethod
+    def _f_2d_nyquist_freq():
+        x_freq_nyquist = 3 + 2
+        y_freq_nyquist = 5 + 3
+        return x_freq_nyquist, y_freq_nyquist
+
+    @pytest.mark.xfail(
+        reason="Numpy, jax, and scipy need to fix bug with 2D FFT (fft2)."
+    )
+    @pytest.mark.unit
+    @pytest.mark.parametrize(
+        "func, m, n",
+        [
+            (_f_2d, 2 * _f_2d_nyquist_freq()[0] + 1, 2 * _f_2d_nyquist_freq()[1] + 1),
+            (_f_2d, 2 * _f_2d_nyquist_freq()[0], 2 * _f_2d_nyquist_freq()[1]),
+        ],
+    )
+    def test_interp_rfft2(self, func, m, n):
```

Comment:
I think this is another fast transform bug in numpy. I'll create a numpy issue and ask soon. The test only passes above the Nyquist rate, when it should pass at the Nyquist rate. The same can reproduced using ``fft2`` instead of ``rfft2``. Does anyone see anything obviously wrong @PlasmaControl/desc-dev ?